widget/view.controller.js

```diff
@@ -0,0 +1,142 @@
+      if ($scope.config.funnelModuleType == 0) { populateData(); }
+      else {
+        populateCustomData();
+      }
+    }
+
+    $scope.color = {
+      layer1 : '#0598A1',
+      layer2 : '#20B4BD',
+      layer3 : '#36CDD7',
+      layer4 : '#3ACAD3'
+    }
+
+    //to populate funnel for custom module
+    function populateCustomData() {
+      $scope.config.moduleList = [];
+      console.log($scope.config.query)
+      var filters = {
+        query: $scope.config.query
+      };
+      var pagedTotalData = new PagedCollection($scope.config.customModule, null, null);
+      pagedTotalData.loadByPost(filters).then(function () {
+        $scope.config.layers.forEach((layer, index) => {
+          var tempArray = layer.value.split('.');
+          if(tempArray.length > 1){
+            var data = pagedTotalData.fieldRows[0][$scope.config.customModuleField].value;
+            tempArray.forEach(function (value){
+              data = data[value];
+            })
+            $scope.config.moduleList.push({ 'title': layer.title, 'data': data })
+          }
+          else{
+            var data = pagedTotalData.fieldRows[0][$scope.config.customModuleField].value[layer.value];
+            $scope.config.moduleList.push({ 'title': layer.title, 'data': data })
+          }
+        });
+        createFunnel();
+      })
+    }
+
+    //populate data for fsr modules
+    function populateData() {
+      var promises = [];
+      $scope.config.moduleList = [];
+      $scope.config.layers.forEach((layer, index) => {
+        var countAggregate = {
+          alias: layer.value,
+          field: '*',
+          operator: 'count'
+        };
+        layer.query.aggregates = [countAggregate];
+        layer.query.limit = ALL_RECORDS_SIZE;
+        var _queryObj = new Query(layer.query);
+        var promise = getResourceData(layer.value, _queryObj).then(function (result) {
+          if (result && result['hydra:member'] && result['hydra:member'].length > 0) {
+            $scope.config.moduleList.push({ 'title': layer.title, 'data': result['hydra:member'][0][layer.value] })
+          }
+        });
+        promises.push(promise);
+      });
+      $q.all(promises).then(function () {
+        createFunnel();
+      })
+    }
+
+    function getResourceData(resource, queryObject) {
+      var defer = $q.defer();
+      $resource(API.QUERY + resource).save(queryObject.getQueryModifiers(), queryObject.getQuery(true)).$promise.then(function (response) {
+        defer.resolve(response);
+      }, function (error) {
+        defer.reject(error);
+      });
+      return defer.promise;
+    }
+
+    function createFunnel() {
+      var margin = 0;
+      var width = 15 + (30 * ($scope.config.layers.length + 3));
+      var red = 30;
+      // var green = 130;
+      // 
+
+      var green = 235 - (($scope.config.layers.length - 1) * 35);
+      var blue = 240;
+      var parentDiv = document.getElementById('parentDiv' + $scope.config.title)
```

Comment:
append unique string to parentDiv id or append wid from config which is unique for every widget